### ISSUE TITLE
Question: Is there a more detailed write-up of the exercise-picking algorithm?

### ISSUE THREAD

**Issue by koljapluemer** [trane-project/trane#288]
Hi,

very cool project!

I'm quite interested in the theory behind this, so I wanted to ask whether there is anything to read on the algorithm handling the Spaced Repetition, interleaving and general exercise picking beyond what's described [here](https://trane-project.github.io/concepts.html) in the documentation?

If not, no worries :)

I'll go back to designing an ukulele curriculum now - thanks for your time!

**Comment by martinmr** [comment]
Hi,

Thanks for trying out Trane. There are more details in the comments of the code, but they might be hard to follow because they are not linear and are interleaved with code.

But the gist of it is that there are two phases:

- In phase one, the graph (or a section of the graph if a filter was specified) is traversed to gather a big set of exercises that the student can answer. The traversal ends when the student does not have a sufficient score to advance to the next courses or lessons.
- In phase two, this first set of exercises is reduced to a smaller batch by first bucketing exercises according to the difficulty and then selecting a portion of the batch from each bucket. This selection is randomized, but is affected by factors such as the score, how deep the exercise is within the graph, how many times it's been selected before, etc. The aim of this phase is to come up with a right mix of exercises.

Relevant source files:
https://github.com/trane-project/trane/blob/master/src/scheduler.rs
https://github.com/trane-project/trane/blob/master/src/scheduler/filter.rs
https://github.com/trane-project/trane/blob/master/src/scorer.rs
